{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dr.Emb Appyter: A Web Platform for Drug Discovery using Embedding Vectors\n",
    "Dr.Emb Appyter is a web-based platform that integrates various embedding methods to search compounds. It searches for closely located compounds of query compounds in the embedding space. <br> <br>\n",
    "The inputs for this Appyter are (a) Library, (b) Query Compound Data (Name, SMILES), and (c) Parameters (top k candidates, embedding method, similarity method). If users want to choose different data or parameters, users can simply go back, make the changes, and then run it again. <br> <br>\n",
    "The default libraries and embedding vectors used to run this Appyter can be found <a href=\"https://github.com/KU-MedAI/The-Dr.Emb-Appyter.git\" target=\"_blank\" rel=\"noopener noreferrer\">here</a>."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import json\n",
    "import requests\n",
    "from tqdm import tqdm\n",
    "import utils\n",
    "\n",
    "import deepchem as dc\n",
    "from rdkit.Chem import AllChem, Lipinski, Descriptors, Crippen\n",
    "from rdkit import Chem, DataStructs\n",
    "import torch\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "import gseapy as gp\n",
    "import methods.moable.model\n",
    "from macaw import *\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import pickle\n",
    "import faiss\n",
    "\n",
    "from upsetplot import plot, from_contents, UpSet\n",
    "from sklearn.manifold import TSNE\n",
    "import umap\n",
    "import plotly.io as po\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import plotly.express as px\n",
    "from itertools import combinations\n",
    "\n",
    "from IPython.display import HTML, display, Markdown, IFrame, FileLink, Image, HTML\n",
    "\n",
    "import certifi\n",
    "import urllib3\n",
    "urllib3.disable_warnings()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "{% do SectionField(\n",
    "    name='LB_Section',\n",
    "    title='Select a Library',\n",
    "    subtitle='Option 1: Select your library: KCB(Korean Chemical Bank, 605K), ZINC(20K), MCE(4.3K), Selleckchem(9K).\n",
    "    Option 2: Upload your custom library.',\n",
    "    img='library.png'\n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Data_Section',\n",
    "    title='Enter Your Query',\n",
    "    subtitle='Load your query as tab separated formats without headers and index columns (One drug per row). \n",
    "    Option 1: One query compound. \n",
    "    Option 2: Multiple query compounds (maximum 10 compounds) as tab-separated formats without headers and index columns (One drug per row)',\n",
    "    img='data.png'\n",
    ") %}\n",
    "\n",
    "{% do SectionField(\n",
    "    name='Parameter_Section',\n",
    "    title='Set Parameters',\n",
    "    subtitle='Top K candidates: the maximum number(K) of top candidates to be returned. \n",
    "        Embedding Methods: Choose an embedding method among ReSimNet, MoAble, Mol2vec, MACAW, ECFP and MACCS Keys. \n",
    "        Similarity Metric: Choose similarity measurement. \n",
    "        We recommend cosine similarity for ReSimNet and MoAble, Euclidean distance for Mol2vec and MACAW and Jaccard similarity for ECFP and MACCS Keys.',\n",
    "    img='setting.png'\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "\n",
    "{% set LB_field = TabField(\n",
    "    name='lb_field',\n",
    "    label='Library Selection',\n",
    "    default='Select a Library',\n",
    "    choices={\n",
    "        'Select a Library': [\n",
    "            ChoiceField(\n",
    "                name='lb_kind',\n",
    "                label='Select your library',\n",
    "                default='MCE',\n",
    "                choices={'KCB':'kcb','ZINC':'zinc','MCE':'mce','Selleckchem':'selleck'}\n",
    "            )\n",
    "        ],\n",
    "        'Custom Library': [\n",
    "            FileField(\n",
    "                name='custom',\n",
    "                label='Load your Custom Library',\n",
    "                default='example_custom_library.tsv',\n",
    "                examples={\"example_custom_library.tsv\":url_for('static', filename='custom_library.tsv')},\n",
    "                description='Upload tab-delimeted file. Columns name should correspond to drug_name, drug_smiles without index columns'),\n",
    "            StringField(\n",
    "                name='custom_info', \n",
    "                label='Custom Information', \n",
    "                default='Description', \n",
    "                description='Columns name of information you want to highlight in the custom library.'\n",
    "            )\n",
    "        ]\n",
    "    },\n",
    "    section = 'LB_Section')\n",
    " %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "{% set input_types = TabField(\n",
    "    name='input_types',\n",
    "    label='Data file',\n",
    "    default='Multiple Query Compounds',\n",
    "    description='Upload your drugs information files',\n",
    "    choices={        \n",
    "        'One Query Compound': [\n",
    "            StringField(\n",
    "                name='drug1_name', \n",
    "                label='Query Compound Name', \n",
    "                default='Tenofovir', \n",
    "                description='', \n",
    "                section='Data_Section'\n",
    "            ),\n",
    "            StringField(\n",
    "                name='drug1_smiles', \n",
    "                label='Query Compound SMILES', \n",
    "                default='C(CN1C=NC2=C(N=CN=C21)N)OCP(=O)(O)O', \n",
    "                description='', \n",
    "                section='Data_Section'\n",
    "            )\n",
    "        ],\n",
    "        'Multiple Query Compounds':[\n",
    "            FileField(\n",
    "                name='drug1_list_filename', \n",
    "                label='List of Candidates (.tsv or .txt)', \n",
    "                default='example_drugs_antiviral.txt', \n",
    "                examples={\"example_drugs_antiviral.txt\":url_for('static', filename='antiviral_drugs.txt')},\n",
    "                description='List of Candidates. Columns should correspond to drug name and drug smiles without any header and index columns. One drug per row',\n",
    "                section='Data_Section'\n",
    "            )]\n",
    "\n",
    "    },\n",
    "    section = 'Data_Section',\n",
    ") %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code_exec\n",
    "\n",
    "{% set topk_candidate = IntField(\n",
    "    name='topk_candidate', \n",
    "    label='Top K Candidates', \n",
    "    min=1, \n",
    "    max=10000, \n",
    "    default=30, \n",
    "    description='The maximum number of Top candidates', \n",
    "    section='Parameter_Section')\n",
    "%}\n",
    "\n",
    "{% set embedding_method = ChoiceField(\n",
    "    name='embedding_method',\n",
    "    label='Embedding Methods',\n",
    "    choices={'ReSimNet (Expression-based)':'ReSimNet','MoAble (Expression-based)':'MoAble','ECFP (Fingerprint-based)':'ECFP','MACCS Keys (Fingerprint-based)':'MACCSKeys',\n",
    "             'Mol2vec (Sequence-based)':'Mol2vec', 'MACAW (Sequence-based)':'MACAW'},\n",
    "    default='ReSimNet (Expression-based)',\n",
    "    description='Select a embedding method',\n",
    "    section='Parameter_Section')\n",
    "%}\n",
    "\n",
    "{% set similarity_method = ChoiceField(\n",
    "    name='similarity_method',\n",
    "    label='Similarity Methods',\n",
    "    choices={'Cosine similarity':'Cosine', 'Euclidean distance':'Euclidean', 'Jaccard similarity':'Jaccard'},\n",
    "    default='Cosine similarity',\n",
    "    description='Select a similarity method to calculate',\n",
    "    section='Parameter_Section')\n",
    "%}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "input_path = 'input_data/'\n",
    "if not os.path.exists(input_path):\n",
    "    os.makedirs(input_path)\n",
    "    \n",
    "output_path = 'output_data/'\n",
    "if not os.path.exists(output_path):\n",
    "    os.makedirs(output_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "# database_field\n",
    "{% if LB_field.raw_value == \"Select a Library\" %}\n",
    "input_db = \"{{ LB_field.value[0] }}\"\n",
    "library_df = utils.library_npl(input_db)[0]\n",
    "npl = utils.library_npl(input_db)[1]\n",
    "\n",
    "{% elif LB_field.raw_value == \"Custom Library\" %}\n",
    "input_db = \"custom\"\n",
    "custom_df = pd.read_csv({{ LB_field.value[0] }}, sep='\\t')\n",
    "label = {{ LB_field.value[1] }}\n",
    "{% endif %}\n",
    "\n",
    "# prototypes_field\n",
    "{% if input_types.raw_value == \"One Query Compound\" %}\n",
    "drug1_name = {{ input_types.value[0] }}\n",
    "drug1_smiles = {{ input_types.value[1] }}\n",
    "input_mode = \"single\"\n",
    "\n",
    "{% elif input_types.raw_value == \"Multiple Query Compounds\" %}\n",
    "query_filename = {{ input_types.value[0] }}\n",
    "input_mode = \"multiple\"\n",
    "{% endif %}\n",
    "\n",
    "# set parameters\n",
    "embed_method = \"{{embedding_method.value}}\"\n",
    "sim_method = \"{{similarity_method.value}}\"\n",
    "topk_candidate = {{topk_candidate.value}}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if similarity_method.value == \"Jaccard\" %}\n",
    "if embed_method not in [\"ECFP\", \"MACCSKeys\"]:\n",
    "    try:\n",
    "        raise ValueError(\"Error: Invalid embed_method. Please use 'ECFP' or 'MACCSKeys' for Embedding Methods.\")\n",
    "    except ValueError as e:\n",
    "        print(e)\n",
    "        assert False    \n",
    "    \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if LB_field.raw_value == \"Custom Library\" %}\n",
    "\n",
    "## Custom Library\n",
    "If users have their own libraries, users can upload a library file to the ‘Custom Library’ field in the input page. The file should contain compound names and SMILES information. <br>\n",
    "Once uploaded, the custom library is preprocessed to generate embedding vectors. <br> <br>\n",
    "Please note that embedding the compounds in a large number of compounds in a custom library may take a long time.\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if LB_field.raw_value == \"Custom Library\" %}\n",
    "\n",
    "library_df, npl = utils.custom_npl(custom_df)\n",
    "custom_dict = dict(zip(library_df[\"drug_name\"], library_df[\"drug_smiles\"]))\n",
    "\n",
    "if embed_method == 'ReSimNet':\n",
    "    input_embed_filename = \"./input_drugs.pkl\"\n",
    "    with open(input_embed_filename, \"wb\") as f:\n",
    "        pickle.dump(custom_dict, f)\n",
    "\n",
    "    resimnet_model = f\"ReSimNet7.mdl\"\n",
    "    custom_output_file = 'Custom_ReSimNet_embedding_vectors_7.pkl'\n",
    "\n",
    "    !python3 ../../../methods/ReSimNet/main.py --save-embed True --embed-d 2 --drug-file $input_embed_filename --output-filename $custom_output_file --checkpoint-dir ../../../methods/ReSimNet/results/ --model-name $resimnet_model --data-path ../../../methods/ReSimNet/tasks/data/drug\\(v0.6\\).pkl\n",
    "\n",
    "    custom_embed_dict = pd.read_pickle(custom_output_file)\n",
    "    \n",
    "    embedding_html = \"<a href=\\\"./{}\\\" target='_blank'>{}</a>\".format(custom_output_file, f\"Download {embed_method} embedding vectors for custom library\")\n",
    "    display(HTML(embedding_html))\n",
    "\n",
    "else:\n",
    "    custom_output_file = f'Custom_{embed_method}_embedding_vectors.pkl'\n",
    "\n",
    "    if embed_method == 'MACAW':\n",
    "        mcw = MACAW()\n",
    "        mcw.fit(library_df['drug_smiles'])\n",
    "        X = mcw.transform(library_df['drug_smiles'])\n",
    "        drug_names = library_df['drug_name']\n",
    "        with open(output_path + custom_output_file, 'wb') as f:\n",
    "            custom_embed_dict = dict(zip(drug_names, X))\n",
    "            pickle.dump(custom_embed_dict, f)\n",
    "    \n",
    "    else:\n",
    "        custom_embed_dict = utils.custom_embedding(custom_dict, embed_method, output_path, custom_output_file)\n",
    "\n",
    "    embedding_html = \"<a href=\\\"./{}\\\" target='_blank'>{}</a>\".format(output_path+custom_output_file, f\"Download {embed_method} embedding vectors for custom library\")\n",
    "    display(HTML(embedding_html))\n",
    "        \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating Embedding Vectors\n",
    "The embedding vectors of the query compounds are generated based on the selected embedding method. Dr.Emb Appyter offers six embedding models: ReSimNet, MoAble, ECFP, MACCS Keys, Mol2vec, and MACAW. <br>\n",
    "The embedding vectors are available for download by clicking the link below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filenames, combined = [], []\n",
    "embed_dict = {}\n",
    "\n",
    "if input_mode == 'multiple':\n",
    "    df = pd.read_csv(query_filename, sep=\"\\t\", header=None)\n",
    "    df.columns = [\"drug1_name\", \"drug1_smiles\"]\n",
    "    drug_dict = dict(zip(df[\"drug1_name\"], df[\"drug1_smiles\"]))\n",
    "    output_file = f'{embed_method}_embedding_vectors.pkl'\n",
    "\n",
    "else:\n",
    "    df = pd.DataFrame([{'drug1_name':drug1_name, 'drug1_smiles':drug1_smiles}])\n",
    "    output_file = f'{embed_method}_embedding_vectors.pkl'\n",
    "    drug_dict = dict(zip(df[\"drug1_name\"], df[\"drug1_smiles\"]))\n",
    "\n",
    "# Gene expression\n",
    "if embed_method == 'ReSimNet':\n",
    "    for name, smiles in drug_dict.items():\n",
    "        globals()[f'input_embed_{name}'] = {name:smiles}\n",
    "        with open(input_path + f'input_{name}.pkl', \"wb\") as f:\n",
    "            pickle.dump(globals()[f'input_embed_{name}'], f)\n",
    "            \n",
    "        input_embed_filename = input_path + f\"input_{name}.pkl\"\n",
    "        output_embed_filename = output_path + f\"output_{name}_embeddings.pkl\"\n",
    "        resimnet_model_filename = \"ReSimNet7.mdl\"\n",
    "        filenames.append(output_embed_filename)\n",
    "        \n",
    "        !python3 ../../../methods/ReSimNet/main.py --save-embed True --embed-d 2 --drug-file $input_embed_filename --output-filename $output_embed_filename --checkpoint-dir ../../../methods/ReSimNet/results/ --model-name $resimnet_model_filename --data-path ../../../methods/ReSimNet/tasks/data/drug\\(v0.6\\).pkl\n",
    "        \n",
    "    for file in filenames:\n",
    "        with open(file, 'rb') as file:\n",
    "            data = pickle.load(file)\n",
    "            combined.append(data)\n",
    "    with open(output_path+output_file, 'wb') as file:\n",
    "        pickle.dump(combined, file)\n",
    "    embed_dict = combined.copy()\n",
    "\n",
    "elif embed_method == 'MoAble':\n",
    "    embed_dict = utils.drug_embeddings(drug_dict)\n",
    "    with open(output_path + output_file,'wb') as f:\n",
    "        pickle.dump(embed_dict, f)\n",
    "    for k, v in embed_dict.items():\n",
    "        with open(output_path + f'{k}_output_embedding.pkl', 'wb') as f:\n",
    "            pickle.dump({k:v}, f)\n",
    "        \n",
    "# Sequence\n",
    "elif embed_method == 'Mol2vec':\n",
    "    featurizer = dc.feat.Mol2VecFingerprint('../../../methods/mol2vec/mol2vec_model_300dim.pkl')\n",
    "    for name, smiles in drug_dict.items():\n",
    "        with open(output_path + f'{name}_output_embedding.pkl', 'wb') as f:\n",
    "            pickle.dump({name: featurizer.featurize(smiles)}, f)\n",
    "        with open(output_path + output_file, 'wb') as f:\n",
    "            globals()[f'input_{name}'] = {name: featurizer.featurize(smiles)}\n",
    "            embed_dict[name] = featurizer.featurize(smiles)\n",
    "            pickle.dump(embed_dict, f)\n",
    "    \n",
    "# Fingerprint\n",
    "elif embed_method == 'ECFP':\n",
    "    for name, smiles in drug_dict.items():\n",
    "        with open(output_path + f'{name}_output_embedding.pkl', 'wb') as f:\n",
    "            pickle.dump({name:utils.smiles2fp(smiles)}, f)\n",
    "        with open(output_path + output_file, 'wb') as f:\n",
    "            globals()[f'input_{name}'] = {name:utils.smiles2fp(smiles)}\n",
    "            embed_dict[name] = utils.smiles2fp(smiles)\n",
    "            pickle.dump(embed_dict, f)\n",
    "\n",
    "\n",
    "elif embed_method == 'MACCSKeys':\n",
    "    featurizer = dc.feat.MACCSKeysFingerprint()\n",
    "    for name, smiles in drug_dict.items():\n",
    "        with open(output_path + f'{name}_output_embedding.pkl', 'wb') as f:\n",
    "            pickle.dump({name:featurizer.featurize(smiles)}, f)\n",
    "        with open(output_path + output_file, 'wb') as f:\n",
    "            globals()[f'input_{name}'] = {name:featurizer.featurize(smiles)}\n",
    "            embed_dict[name] = featurizer.featurize(smiles)\n",
    "            for drug, values in embed_dict.items():\n",
    "                embed_dict[drug] = values.flatten()\n",
    "            pickle.dump(embed_dict, f)\n",
    "\n",
    "\n",
    "elif embed_method == 'MACAW':\n",
    "    if input_db != 'custom':\n",
    "        mcw = utils.pretrained_MACAW(input_db)\n",
    "    for name, smiles in drug_dict.items():\n",
    "        with open(output_path + f'{name}_output_embedding.pkl', 'wb') as f:\n",
    "            pickle.dump({name:mcw.transform([smiles])}, f)\n",
    "        with open(output_path + output_file, 'wb') as f:\n",
    "            globals()[f'input_{name}'] = {name:mcw.transform([smiles])}\n",
    "            embed_dict[name] = mcw.transform([smiles])\n",
    "            pickle.dump(embed_dict, f)\n",
    "\n",
    "embedding_html = \"<a href=\\\"./{}\\\" target='_blank'>{}</a>\".format(output_path+output_file, f\"Download {embed_method} embedding vectors for query compounds\")\n",
    "display(HTML(embedding_html))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compounds Search\n",
    "Dr.Emb Appyter uses the Faiss-based search system to calculate and compare distances between query embedding vectors and library embedding vectors. Faiss efficiently measures the distances between the library and query compounds and returns the top closest compounds as a result. <br> <br>\n",
    "The distance between embedding vectors is calculated based on the selected similarity method. Please note that Jaccard similarity is calculated only for binary vectors such as ECFP and MACCS Keys."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# custom_result_df\n",
    "if sim_method != 'Jaccard':\n",
    "    for name, smiles in drug_dict.items():\n",
    "        data_filename = f\"{name}_candidates.tsv\"\n",
    "        if input_db == 'custom':\n",
    "            if embed_method == 'ReSimNet':\n",
    "                globals()[f'{name}_simliarity'], globals()[f'{name}_index'], globals()[f'{name}_result_df'] = utils.custom_finder(custom_embed_dict, embed_method, npl, sim_method, output_path+f'output_{name}_embeddings.pkl', topk_candidate, name)\n",
    "            else:\n",
    "                globals()[f'{name}_simliarity'], globals()[f'{name}_index'], globals()[f'{name}_result_df'] = utils.custom_finder(custom_embed_dict, embed_method, npl, sim_method, output_path+f'{name}_output_embedding.pkl', topk_candidate, name)\n",
    "\n",
    "            globals()[f'{name}_result_df'] = pd.concat(globals()[f'{name}_result_df'], axis=1).reset_index()\n",
    "            globals()[f'{name}_result_df'] = globals()[f'{name}_result_df'].rename(columns={\"index\":\"drug_name\" if input_db == 'custom' else 'drug2_name'})\n",
    "            globals()[f'{name}_result_df'] = pd.merge(globals()[f'{name}_result_df'], library_df)\n",
    "            globals()[f'{name}_result_df'][f'{sim_method} similarity'] = globals()[f'{name}_simliarity'][0]\n",
    "            globals()[f'{name}_result_df']['rank'] = [x for x in range(1,topk_candidate+1)]\n",
    "            globals()[f'{name}_result_df'] = globals()[f'{name}_result_df'].set_index(['rank'])\n",
    "            if input_db == 'custom':\n",
    "                globals()[f'{name}_result_df']['drug_smiles'] = globals()[f'{name}_result_df'].apply(lambda x: utils.make_clickable(x['drug_smiles']), axis=1)\n",
    "                globals()[f'{name}_result_df'].rename(columns={'drug_name':'compound_name', 'drug_smiles':'compound_smiles'}, inplace=True)\n",
    "            else:\n",
    "                globals()[f'{name}_result_df']['drug2_smiles'] = globals()[f'{name}_result_df'].apply(lambda x: utils.make_clickable(x['drug2_smiles']), axis=1)\n",
    "                globals()[f'{name}_result_df'].rename(columns={'drug2_name':'compound_name', 'drug2_smiles':'compound_smiles'}, inplace=True)\n",
    "            globals()[f'{name}_result_df'].to_csv(data_filename, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if sim_method == 'Jaccard':\n",
    "    data_filename = f\"{name}_candidates.tsv\"\n",
    "    query_dataframes = utils.jaccard_dataframes(input_db, custom_embed_dict if input_db=='custom' else embed_dict, embed_method, embed_dict, topk_candidate)\n",
    "    for query_name, df in query_dataframes.items():\n",
    "        globals()[f'{query_name}_result_df'] = df\n",
    "        if input_db != 'custom':\n",
    "            globals()[f'{query_name}_result_df'].rename(columns={'drug_name':'drug2_name'}, inplace=True)\n",
    "        globals()[f'{query_name}_result_df'] = pd.merge(globals()[f'{query_name}_result_df'], library_df)\n",
    "        second_column = globals()[f'{query_name}_result_df'].pop('Jaccard Similarity')\n",
    "        globals()[f'{query_name}_result_df']['Jaccard Similarity'] = second_column\n",
    "        globals()[f'{query_name}_result_df']['rank'] = [x for x in range(1,topk_candidate+1)]\n",
    "        globals()[f'{query_name}_result_df'] = globals()[f'{query_name}_result_df'].set_index(['rank'])\n",
    "        if input_db == 'custom':\n",
    "            globals()[f'{query_name}_result_df']['drug_smiles'] = globals()[f'{query_name}_result_df'].apply(lambda x: utils.make_clickable(x['drug_smiles']), axis=1)\n",
    "            globals()[f'{query_name}_result_df'].rename(columns={'drug_name':'compound_name', 'drug_smiles':'compound_smiles'}, inplace=True)\n",
    "        else:\n",
    "            globals()[f'{query_name}_result_df']['drug2_smiles'] = globals()[f'{query_name}_result_df'].apply(lambda x: utils.make_clickable(x['drug2_smiles']), axis=1)\n",
    "            globals()[f'{query_name}_result_df'].rename(columns={'drug2_name':'compound_name', 'drug2_smiles':'compound_smiles'}, inplace=True)\n",
    "        globals()[f'{query_name}_result_df'].to_csv(data_filename, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# result_df\n",
    "if sim_method != 'Jaccard':\n",
    "    for name, smiles in drug_dict.items():\n",
    "        if input_db != 'custom':\n",
    "            if embed_method == 'ReSimNet':\n",
    "                globals()[f'{name}_simliarity'], globals()[f'{name}_index'], globals()[f'{name}_result_df'] = utils.resimnet_finder(input_db, npl, output_path+f'output_{name}_embeddings.pkl', topk_candidate, name, \"ReSimNet7\")\n",
    "\n",
    "                globals()[f'{name}_result_df'] = pd.concat(globals()[f'{name}_result_df'], axis=1).reset_index()\n",
    "                globals()[f'{name}_result_df'] = globals()[f'{name}_result_df'].rename(columns={\"index\":\"drug2_name\"})\n",
    "                globals()[f'{name}_result_df'] = pd.merge(globals()[f'{name}_result_df'], library_df)\n",
    "                \n",
    "                globals()[f'{name}_df'] = globals()[f'{name}_result_df'].copy()\n",
    "                globals()[f'{name}_df'][\"drug1_name\"] = name\n",
    "                globals()[f'{name}_df'][\"drug1_smiles\"] = smiles\n",
    "\n",
    "                input_filename = f\"./input_pairs_{name}.tsv\"\n",
    "                output_filename = f\"output_data/output_pairs_{name}.tsv\"\n",
    "                globals()[f'{name}_df'].to_csv(input_filename, index=None, sep=\"\\t\")\n",
    "\n",
    "                # resimnet score calculation\n",
    "                !python3 ../../../methods/ReSimNet/main.py --save-prediction-new-pairs True --new-drug-pair-filename $input_filename --output-filename $output_filename --checkpoint-dir ../../../methods/ReSimNet/results/ --data-path ../../../methods/ReSimNet/tasks/data/drug\\(v0.6\\).pkl\n",
    "            \n",
    "            else:\n",
    "                if embed_method in ['MoAble', 'ECFP']:\n",
    "                    globals()[f'{name}_simliarity'], globals()[f'{name}_index'], globals()[f'{name}_result_df'] = utils.MA_finder(input_db, embed_method, npl, sim_method, output_path+f'{name}_output_embedding.pkl', topk_candidate, name)\n",
    "                else:\n",
    "                    globals()[f'{name}_simliarity'], globals()[f'{name}_index'], globals()[f'{name}_result_df'] = utils.finder(input_db, embed_method, npl, sim_method, output_path+f'{name}_output_embedding.pkl', topk_candidate, name)\n",
    "\n",
    "                globals()[f'{name}_result_df'] = pd.concat(globals()[f'{name}_result_df'], axis=1).reset_index()\n",
    "                globals()[f'{name}_result_df'] = globals()[f'{name}_result_df'].rename(columns={\"index\":\"drug_name\" if input_db == 'custom' else 'drug2_name'})\n",
    "                globals()[f'{name}_result_df'] = pd.merge(globals()[f'{name}_result_df'], library_df)\n",
    "                globals()[f'{name}_result_df'][f'{sim_method} similarity' if sim_method!='Euclidean' else f'{sim_method} distance'] = globals()[f'{name}_simliarity'][0]\n",
    "                globals()[f'{name}_result_df']['rank'] = [x for x in range(1,topk_candidate+1)]\n",
    "                globals()[f'{name}_result_df'] = globals()[f'{name}_result_df'].set_index(['rank'])\n",
    "                if input_db == 'custom':\n",
    "                    globals()[f'{name}_result_df']['drug_smiles'] = globals()[f'{name}_result_df'].apply(lambda x: utils.make_clickable(x['drug_smiles']), axis=1)\n",
    "                    globals()[f'{name}_result_df'].rename(columns={'drug_name':'compound_name', 'drug_smiles':'compound_smiles'}, inplace=True)\n",
    "                else:\n",
    "                    globals()[f'{name}_result_df']['drug2_smiles'] = globals()[f'{name}_result_df'].apply(lambda x: utils.make_clickable(x['drug2_smiles']), axis=1)\n",
    "                    globals()[f'{name}_result_df'].rename(columns={'drug2_name':'compound_name', 'drug2_smiles':'compound_smiles'}, inplace=True)\n",
    "                globals()[f'{name}_result_df'].to_csv(data_filename, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if sim_method != 'Jaccard':\n",
    "    if input_db != 'custom' and embed_method == 'ReSimNet':\n",
    "        for name, smiles in drug_dict.items():\n",
    "            globals()[f'{name}_result_df'] = pd.read_csv(output_path + f'output_pairs_{name}.tsv', sep=\"\\t\")\n",
    "            globals()[f'{name}_result_df']['drug2_smiles'] = globals()[f'{name}_result_df'].apply(lambda x: utils.make_clickable(x['drug2_smiles']), axis=1)\n",
    "            globals()[f'{name}_result_df'] = globals()[f'{name}_result_df'].set_index(['drug2_name', 'drug2_smiles'])\n",
    "            globals()[f'{name}_result_df'] = globals()[f'{name}_result_df'].drop(['drug1_name','drug1_smiles'], axis=1)\n",
    "            globals()[f'{name}_result_df'][\"ReSimNet_avg\"] = globals()[f'{name}_result_df'][[x for x in globals()[f'{name}_result_df'].columns if x.startswith(\"ReSimNet\")]].mean(axis=1)\n",
    "            globals()[f'{name}_result_df']['ReSimNet_std'] = globals()[f'{name}_result_df'][[x for x in globals()[f'{name}_result_df'].columns if x.startswith('ReSimNet')]].std(axis=1)\n",
    "            # order columns\n",
    "            globals()[f'{name}_result_df'] = globals()[f'{name}_result_df'].reindex(sorted(globals()[f'{name}_result_df'].columns), axis=1)\n",
    "            globals()[f'{name}_result_df'] = globals()[f'{name}_result_df'].sort_values([\"ReSimNet_avg\", \"jaccard_similarity\"], ascending=False)\n",
    "            globals()[f'{name}_result_df'].reset_index(drop=False, inplace=True)\n",
    "            # rank\n",
    "            globals()[f'{name}_result_df']['rank'] = [x for x in range(1,topk_candidate+1)]\n",
    "            globals()[f'{name}_result_df'] = globals()[f'{name}_result_df'].set_index(['rank'])\n",
    "            globals()[f'{name}_result_df'].rename(columns={'drug2_name':'compound_name', 'drug2_smiles':'compound_smiles'}, inplace=True)\n",
    "            \n",
    "            # html\n",
    "            globals()[f'{name}_result_df1'] = globals()[f'{name}_result_df'].copy() \n",
    "            globals()[f'{name}_result_df1'].drop(columns=['ReSimNet_0','ReSimNet_1','ReSimNet_2','ReSimNet_3','ReSimNet_4','ReSimNet_5','ReSimNet_6','ReSimNet_7','ReSimNet_8','ReSimNet_9'], inplace=True)\n",
    "            globals()[f'{name}_result_df1'].rename(columns={'drug2_name':'compound_name', 'drug2_smiles':'compound_smiles'}, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Search Results\n",
    "The search results for the query compound(s) are presented as follows:\n",
    "-\tThe top k compounds are retrieved from the library and ranked based on the highest similarity.\n",
    "-\tThe names and SMILES of the compounds in the library are displayed in the 'compound_name' and 'compound_smiles' columns, respectively.\n",
    "-\tThe selected similarity (or distance) results are shown in the similarity (or distance) column.\n",
    "-\tFor custom libraries, user-specified column information is displayed accordingly in the result table.\n",
    "-\tAdditionally, the SMILES information is interactive, with clickable links redirecting to PubChem.\n",
    "-\tThe results can be downloaded as a .tsv (tab-separated values) file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "css = \"\"\"\n",
    "<style>\n",
    "table.dataframe td{\n",
    "    white-space: pre-wrap;\n",
    "    max-width: 450px;\n",
    "    overflow: hidden;\n",
    "}\n",
    "</style>\n",
    "\"\"\"\n",
    "\n",
    "tb_number = 1\n",
    "styles = [dict(selector=\"caption\", props=[(\"text-align\", \"left\")])]\n",
    "for name in drug_dict.keys():\n",
    "    if input_db != 'custom' and embed_method == 'ReSimNet':\n",
    "        caption = f\"Table {tb_number}. Top {topk_candidate} compounds similar to {name} based on {embed_method} embedding\"\n",
    "        globals()[f'{name}_result_df'].to_csv(f'{name}_results.tsv',sep='\\t')\n",
    "        tb_number += 1\n",
    "        globals()[f'{name}_html'] = \"<a href=\\\"./{}\\\" target='_blank'>{}</a>\".format(f'{name}_results.tsv', \"Download results: {}\".format(f'{name}_results.tsv'))\n",
    "        html_table = css+globals()[f'{name}_result_df1'].to_html(classes='styled-table', escape=False)\n",
    "        styled_html = f\"<caption>{caption}</caption>{html_table}\"\n",
    "        result_html = \"<a href=\\\"./{}\\\" target='_blank'>{}</a>\".format(f'{name}_results.tsv', \"Download results: {}\".format(f'{name}_results.tsv'))\n",
    "        globals()[f'{name}_html'] = \"<a href=\\\"./{}\\\" target='_blank'>{}</a>\".format(f'{name}_results.tsv', \"Download results: {}\".format(f'{name}_results.tsv'))\n",
    "        display(HTML(styled_html))\n",
    "        display(HTML(result_html))\n",
    "        \n",
    "    else:\n",
    "        globals()[f'{name}_result_df1'] = globals()[f'{name}_result_df'].copy()\n",
    "        caption = f\"Table {tb_number}. Top {topk_candidate} compounds similar to {name} based on {embed_method} embedding\"\n",
    "        globals()[f'{name}_result_df'].to_csv(f'{name}_results.tsv',sep='\\t')\n",
    "        tb_number += 1\n",
    "        globals()[f'{name}_html'] = \"<a href=\\\"./{}\\\" target='_blank'>{}</a>\".format(f'{name}_results.tsv', \"Download results: {}\".format(f'{name}_results.tsv'))\n",
    "        html_table = css+globals()[f'{name}_result_df1'].to_html(classes='styled-table', escape=False)\n",
    "        styled_html = f\"<caption>{caption}</caption>{html_table}\"\n",
    "        result_html = \"<a href=\\\"./{}\\\" target='_blank'>{}</a>\".format(f'{name}_results.tsv', \"Download results: {}\".format(f'{name}_results.tsv'))\n",
    "        globals()[f'{name}_html'] = \"<a href=\\\"./{}\\\" target='_blank'>{}</a>\".format(f'{name}_results.tsv', \"Download results: {}\".format(f'{name}_results.tsv'))\n",
    "        display(HTML(styled_html))\n",
    "        display(HTML(result_html))\n",
    "\n",
    "\n",
    "    topk_embedding_file = f'{name}_top_{topk_candidate}_embedding_vectors.pkl'\n",
    "    if input_db == 'custom':\n",
    "        globals()[f'{name}_final_vectors'] = {key: custom_embed_dict[key] for key in globals()[f'{name}_result_df']['compound_name'] if key in custom_embed_dict}\n",
    "    else:\n",
    "        lib_embed_dict = utils.embed_vector_lib(input_db, embed_method)\n",
    "        globals()[f'{name}_final_vectors'] = {key: lib_embed_dict[key] for key in globals()[f'{name}_result_df']['compound_name'] if key in lib_embed_dict}\n",
    "\n",
    "    with open(output_path+topk_embedding_file, 'wb') as f:\n",
    "        pickle.dump(globals()[f'{name}_final_vectors'], f)\n",
    "        \n",
    "    embedding_html = \"<a href=\\\"./{}\\\" target='_blank'>{}</a>\".format(output_path+topk_embedding_file, f\"Download results: {name} Top {topk_candidate} {embed_method} embedding vectors\")\n",
    "    display(HTML(embedding_html))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if input_types.raw_value == \"Multiple Query Compounds\" %}\n",
    "\n",
    "### Overlapping Compounds Information\n",
    "Dr.Emb Appyter checks for any overlapping compounds between the top k compounds of query compounds and displays their information in a table. <br> <br>\n",
    "If there are overlapping compounds, their names, SMILES, and the query compounds they overlap with are displayed in a table. The column names are compound_name, compound_smiles, and overlap, respectively. However, if no overlapping compounds are detected, the table is omitted, and the message \"There are no overlapping compounds.\" is displayed.\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if input_types.raw_value == \"Multiple Query Compounds\" %}\n",
    "\n",
    "upset_dict = {}\n",
    "rdf = []\n",
    "\n",
    "for name in drug_dict.keys():\n",
    "    upset_dict[f\"{name}\"] = list(globals()[f'{name}_result_df']['compound_name'])\n",
    "\n",
    "result_keys = utils.find_duplicate_names(upset_dict)\n",
    "\n",
    "for compound_name, values in result_keys.items():\n",
    "    if compound_name in library_df['drug2_name' if input_db != \"custom\" else \"drug_name\"].values:\n",
    "        compound_smiles = library_df.loc[library_df['drug2_name' if input_db != \"custom\" else \"drug_name\"] == compound_name, 'drug2_smiles' if input_db != \"custom\" else \"drug_smiles\"].values[0]\n",
    "        rdf.append({'compound_name': compound_name, 'compound_smiles': compound_smiles, 'overlap': values})\n",
    "\n",
    "ovlap_df = pd.DataFrame(rdf)\n",
    "\n",
    "if not ovlap_df.empty:\n",
    "    caption = f\"Table {tb_number}. Overlapping compounds results based on {embed_method} embedding\"\n",
    "    tb_number += 1\n",
    "    ovlap_df['compound_smiles'] = ovlap_df.apply(lambda x: utils.make_clickable(x['compound_smiles']), axis=1)\n",
    "    ovlap_table = ovlap_df.to_html(classes='styled-table', escape=False)\n",
    "    ovlap_df_html = f\"<caption>{caption}</caption>{ovlap_table}\"\n",
    "    ovlap_html = \"<a href=\\\"./{}\\\" target='_blank'>{}</a>\".format('overlap_results.tsv', \"Download results: {}\".format(f'overlap_results.tsv'))\n",
    "\n",
    "    display(HTML(ovlap_df_html))\n",
    "    display(HTML(ovlap_html))\n",
    "else:\n",
    "    print(\"There are no overlapping compounds.\")\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scatter Plot\n",
    "A scatter plot by UMAP visualizes all the compounds in the library (denoted as L, in gray, with a circle shape), along with the query compounds (Q, diamond) and the top k compounds (Q_topk, circle) in the embedding space. <br> <br>\n",
    "-\tNode size: Nodes increase in size with the number of overlapping compounds. The default node size is 0.3, the size of query compound nodes is 0.5, and the size of overlapping compounds increases incrementally by 1.\n",
    "-\tColor: The default node color is gray. For custom libraries, node color varies based on the user-specified column. The top k compounds for each query compound are distinctively colored.\n",
    "-\tShape: Query compounds are represented by diamond-shaped nodes, while the top k compounds and the library compounds are shown as circles.\n",
    "-\tHover Information: Hovering over a node reveals details such as compound name, SMILES, coordinates of the compound embedding vector, node size, and overlapping query compounds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if input_db == 'custom':\n",
    "    library_dr = custom_embed_dict\n",
    "else:\n",
    "    if embed_method == 'ReSimNet':\n",
    "        library_dr = pd.read_pickle(f'../../../Library/ReSimNet_{input_db}/ReSimNet_{input_db}_7.pkl')\n",
    "    elif embed_method == 'MACCSKeys':\n",
    "        library_dr = pd.read_pickle(f'../../../Library/{embed_method}_{input_db}/{embed_method}_{input_db}.pkl')\n",
    "        for drug, values in library_dr.items():\n",
    "            library_dr[drug] = values.flatten()      \n",
    "    else:\n",
    "        library_dr = pd.read_pickle(f'../../../Library/{embed_method}_{input_db}/{embed_method}_{input_db}.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if input_db == 'custom':\n",
    "    if embed_method in ['Mol2vec', 'MACAW']:\n",
    "        data_array_l = np.array([v[0] for v in custom_embed_dict.values()])\n",
    "        L = pd.DataFrame(data_array_l, columns=[i for i in range(data_array_l.shape[1])], index=custom_embed_dict.keys())\n",
    "        data_array_q = np.array([v[0] for v in embed_dict.values()])\n",
    "        Q = pd.DataFrame(data_array_q, columns=[i for i in range(data_array_q.shape[1])], index=embed_dict.keys())\n",
    "        \n",
    "    elif embed_method in ['MoAble', 'ECFP', 'MACCSKeys']:\n",
    "        L = pd.DataFrame.from_dict(library_dr).T\n",
    "        Q = pd.DataFrame.from_dict(embed_dict).T\n",
    "    \n",
    "    elif embed_method in ['ReSimNet']:\n",
    "        L = pd.DataFrame.from_dict(library_dr).T\n",
    "        dataframes = []\n",
    "        for d in embed_dict:\n",
    "            for key, value in d.items():\n",
    "                dataframes.append(pd.DataFrame({key: value}))\n",
    "        Q = pd.concat(dataframes, axis=1).T\n",
    "    \n",
    "else:        \n",
    "    if embed_method in ['Mol2vec', 'MACAW']:\n",
    "        data_array_l = np.array([v[0] for v in library_dr.values()])\n",
    "        L = pd.DataFrame(data_array_l, columns=[i for i in range(data_array_l.shape[1])], index=library_dr.keys())\n",
    "        data_array_q = np.array([v[0] for v in embed_dict.values()])\n",
    "        Q = pd.DataFrame(data_array_q, columns=[i for i in range(data_array_q.shape[1])], index=embed_dict.keys())\n",
    "        \n",
    "    elif embed_method == 'ReSimNet':\n",
    "        L = pd.DataFrame.from_dict(library_dr).T\n",
    "        dataframes = []\n",
    "        for d in embed_dict:\n",
    "            for key, value in d.items():\n",
    "                dataframes.append(pd.DataFrame({key: value}))\n",
    "        Q = pd.concat(dataframes, axis=1).T\n",
    "        \n",
    "    else:\n",
    "        L = pd.DataFrame.from_dict(library_dr).T\n",
    "        Q = pd.DataFrame.from_dict(embed_dict).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "L.reset_index(inplace=True)\n",
    "Q.reset_index(inplace=True)\n",
    "L.rename(columns={'index':'compound_name'}, inplace=True)\n",
    "Q.rename(columns={'index':'compound_name'}, inplace=True)\n",
    "\n",
    "# custom\n",
    "Q['node_shape'] = Q['compound_name']\n",
    "L['node_shape'] = f'{input_db}'\n",
    "Q['node_size'] = 0.5\n",
    "L['node_size'] = 0.3\n",
    "\n",
    "if input_db == 'custom':\n",
    "    Q['color'] = Q['compound_name']\n",
    "    L['color'] = library_df[label]\n",
    "    \n",
    "all_labels = []\n",
    "for name in drug_dict.keys():\n",
    "    globals()[f'{name}_toplist'] = list(globals()[f'{name}_result_df']['compound_name'])\n",
    "    L.loc[L['compound_name'].isin(globals()[f'{name}_toplist']), 'node_shape'] = f'{name}_top{topk_candidate}'\n",
    "    all_labels.append(globals()[f'{name}_toplist'])\n",
    "all_top_list = [item for sublist in all_labels for item in sublist]\n",
    "label_counts = {label: all_top_list.count(label) for label in set(all_top_list)}\n",
    "data = {'compound_name': list(label_counts.keys()), 'node_size': list(label_counts.values())}\n",
    "dataframe = pd.DataFrame(data)\n",
    "dataframe.loc[dataframe['node_size'] == 1, 'node_size'] = 0.3\n",
    "dataframe.set_index('compound_name', inplace=True)\n",
    "L.set_index('compound_name', inplace=True)\n",
    "L.update(dataframe)\n",
    "L.reset_index(inplace=True)\n",
    "\n",
    "library_query = pd.concat([L, Q], axis=0, ignore_index=True)\n",
    "\n",
    "if embed_method in ['ReSimNet', 'Mol2vec']:\n",
    "    features = library_query.loc[:,0:299]\n",
    "elif embed_method in ['ECFP']:\n",
    "    features = library_query.loc[:,0:2047]\n",
    "elif embed_method in ['MACAW']:\n",
    "    features = library_query.loc[:,0:14]\n",
    "elif embed_method in ['MACCSKeys']:\n",
    "    features = library_query.loc[:,0:166]\n",
    "else:\n",
    "    features = library_query.loc[:,0:255]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if input_types.raw_value == \"Multiple Query Compounds\" %}\n",
    "\n",
    "library_query['Overlap'] = \"-\"\n",
    "\n",
    "for index, row in library_query.iterrows():\n",
    "    compound_name = row['compound_name']\n",
    "    if compound_name in result_keys:\n",
    "        substances = result_keys[compound_name]\n",
    "        library_query.at[index, 'Overlap'] = ', '.join(substances)\n",
    "        \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Umap = umap.UMAP(n_components=3, metric='cosine')\n",
    "umap_projections = Umap.fit_transform(features, )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "special_labels = [label for label in library_query['node_shape'].unique() if not (f'{input_db}' in label or label.endswith(f'_top{topk_candidate}'))]\n",
    "color_discrete_map = {f'{input_db}': 'gray'}\n",
    "hover_data_config = {}\n",
    "\n",
    "if 'Overlap' in library_query.columns:\n",
    "    hover_data_config['Overlap'] = library_query.Overlap.tolist()\n",
    "\n",
    "if input_db == 'custom':\n",
    "    fig = px.scatter_3d(\n",
    "        umap_projections, x=0, y=1, z=2,\n",
    "        color=library_query.color,\n",
    "        labels={'color': 'label'},\n",
    "        color_discrete_map=color_discrete_map,\n",
    "        hover_name=library_query['compound_name'],\n",
    "        size=library_query.node_size,\n",
    "        opacity=0.3,\n",
    "        hover_data=hover_data_config\n",
    "    )\n",
    "    \n",
    "else:\n",
    "    fig = px.scatter_3d(\n",
    "        umap_projections, x=0, y=1, z=2,\n",
    "        color=library_query.node_shape,\n",
    "        labels={'color': 'label'},\n",
    "        color_discrete_map=color_discrete_map,\n",
    "        hover_name=library_query['compound_name'],\n",
    "        size=library_query.node_size,\n",
    "        opacity=0.3,\n",
    "        hover_data=hover_data_config\n",
    "    )\n",
    "    \n",
    "fig.update_traces(marker=dict(line=dict(width=0)))\n",
    "\n",
    "fig.update_layout(\n",
    "    width=800,\n",
    "    height=600,\n",
    "    title=f'Scatter plot of {input_db}_UMAP'\n",
    ")\n",
    "\n",
    "for trace in fig.data:\n",
    "    if trace.name in special_labels:\n",
    "        trace.marker.symbol = 'diamond'\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if input_types.raw_value == \"Multiple Query Compounds\" %}\n",
    "\n",
    "### UpSet Plot\n",
    "In this section, Dr.Emb Appyter visualizes overlapping compounds graphically. An Upset plot illustrates any overlapping compounds found for each query compound. <br> <br>\n",
    "When there are overlapping compounds, the black dots under the query compounds in the plot are connected. These connections indicate that the compounds have shared results. <br> <br>\n",
    "If there are no overlapping compounds, the black dots are not connected, and the bar graph represents the number of compounds.\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if input_types.raw_value == \"Multiple Query Compounds\" %}\n",
    "\n",
    "upsetplot = from_contents(upset_dict)\n",
    "UpSet(upsetplot, subset_size='count', show_counts=True).plot()\n",
    "plt.title('UpSet Plot')\n",
    "plt.savefig('Upset_plot.png')\n",
    "plt.figtext(0.5, 0.01, f\"figure 1. UpSet Plot for the Entered Query Compounds Results\", ha=\"center\", fontsize=10)\n",
    "display(HTML(f'<a href=\"Upset_plot.png\" download>Download: UpSet Plot</a>'))\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if input_types.raw_value == \"Multiple Query Compounds\" %}\n",
    "\n",
    "### Heatmap\n",
    "Dr.Emb Appyter shows similarities between user-queried compounds. The heatmap visualizes the similarity between the embedding vectors of the query compounds entered by the user, facilitating an assessment of the likelihood that the query compounds share similar properties.\n",
    "<br> <br>\n",
    "Please note that ReSimNet produces two heatmaps based on ReSimNet score and Jaccard similarity.\n",
    "\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if input_types.raw_value == \"Multiple Query Compounds\" %}\n",
    "if embed_method == 'ReSimNet':\n",
    "    df = pd.read_csv(query_filename, sep=\"\\t\", header=None)\n",
    "    df.columns = [\"drug2_name\", \"drug2_smiles\"]\n",
    "\n",
    "    for name, smiles in drug_dict.items():\n",
    "        globals()[f'input_{name}_df'] = df\n",
    "        globals()[f'input_{name}_df']['drug1_name'] = name\n",
    "        globals()[f'input_{name}_df']['drug1_smiles'] = smiles\n",
    "        \n",
    "        input_filename = f\"./input_heatmap_{name}.tsv\"\n",
    "        output_filename = f\"./output_heatmap_{name}.tsv\"\n",
    "        globals()[f'input_{name}_df'].to_csv(input_filename, index=None, sep=\"\\t\")\n",
    "        \n",
    "        !python3 ../../../methods/ReSimNet/main.py --save-prediction-new-pairs True --new-drug-pair-filename $input_filename --output-filename $output_filename --checkpoint-dir ../../../methods/ReSimNet/results/ --data-path ../../../methods/ReSimNet/tasks/data/drug\\(v0.6\\).pkl\n",
    "\n",
    "else:\n",
    "    drug_names = list(embed_dict.keys())\n",
    "    if sim_method == 'Cosine':\n",
    "        vectors = np.array(list(embed_dict.values()))\n",
    "        vectors = vectors.squeeze()\n",
    "        cosine_similarity_matrix = utils.calculate_cosine_similarity(vectors)\n",
    "        df = pd.DataFrame(cosine_similarity_matrix, index=drug_names, columns=drug_names)\n",
    "        \n",
    "    elif sim_method == 'Euclidean':\n",
    "        num_drugs = len(drug_names)\n",
    "        df = np.zeros((num_drugs, num_drugs))\n",
    "        for i in range(num_drugs):\n",
    "            for j in range(num_drugs):\n",
    "                if i != j:\n",
    "                    df[i, j] = utils.euclidean_distance(embed_dict[drug_names[i]], embed_dict[drug_names[j]])\n",
    "\n",
    "    elif sim_method == 'Jaccard':\n",
    "        embed_df = pd.DataFrame(embed_dict)\n",
    "        df = pd.DataFrame(np.nan, index=embed_df.columns, columns=embed_df.columns)\n",
    "        for drug1, drug2 in combinations(embed_df.columns, 2):\n",
    "            similarity = utils.jaccard_similarity(embed_df[drug1], embed_df[drug2])\n",
    "            df.at[drug1, drug2] = similarity\n",
    "            df.at[drug2, drug1] = similarity\n",
    "        np.fill_diagonal(df.values, 1)\n",
    "\n",
    "    labels = list(embed_dict.keys())\n",
    "    df = pd.DataFrame(df, index=labels, columns=labels)\n",
    "    sns.set(font_scale=1.0)\n",
    "    plt.figure(figsize=(6, 6))\n",
    "    sns.heatmap(df, cmap=\"OrRd\", annot=True, fmt=\".2f\", cbar_kws={'label': f'{sim_method}'})\n",
    "    plt.title(f\"{embed_method}_{sim_method}\")\n",
    "    plt.savefig(f'{embed_method}_{sim_method}.png')\n",
    "    plt.figtext(0.5, 0.01, f\"figure 2. Heatmap for the Entered Query Compounds\", ha=\"center\", fontsize=10)\n",
    "    plt.show()\n",
    "    display(HTML(f'<a href=\"./{embed_method}_{sim_method}.png\" download>Download: {embed_method}_{sim_method} heatmap</a>')) \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if input_types.raw_value == \"Multiple Query Compounds\" %}\n",
    "data_avg = []\n",
    "data_jac = []\n",
    "if embed_method == 'ReSimNet':\n",
    "    for name in drug_dict.keys():\n",
    "        output_filename = f'./output_heatmap_{name}.tsv'\n",
    "        \n",
    "        globals()[f'{name}_result_hp'] = pd.read_csv(output_filename, sep=\"\\t\")\n",
    "        globals()[f'{name}_result_hp']['drug2_smiles'] = globals()[f'{name}_result_hp'].apply(lambda x: utils.make_clickable(x['drug2_smiles']), axis=1)\n",
    "        globals()[f'{name}_result_hp'][\"ReSimNet_avg\"] = globals()[f'{name}_result_hp'][[x for x in globals()[f'{name}_result_hp'].columns if x.startswith(\"ReSimNet\")]].mean(axis=1)\n",
    "        globals()[f'{name}_result_hp_avg'] = globals()[f'{name}_result_hp'][['drug1_name','drug2_name','ReSimNet_avg']]\n",
    "        globals()[f'{name}_result_hp_jac'] = globals()[f'{name}_result_hp'][['drug1_name','drug2_name','jaccard_similarity']]\n",
    "        globals()[f'{name}_result_hp_avg_pv'] = globals()[f'{name}_result_hp_avg'].pivot(index='drug1_name', columns='drug2_name', values='ReSimNet_avg')\n",
    "        globals()[f'{name}_result_hp_jac_pv'] = globals()[f'{name}_result_hp_jac'].pivot(index='drug1_name', columns='drug2_name', values='jaccard_similarity')\n",
    "        \n",
    "        data_avg.append(globals()[f'{name}_result_hp_avg_pv'])\n",
    "        data_jac.append(globals()[f'{name}_result_hp_jac_pv'])\n",
    "        heatmap_avg = pd.concat(data_avg, axis=0)\n",
    "        heatmap_avg = heatmap_avg.sort_index(axis=0).sort_index(axis=1)\n",
    "        heatmap_jac = pd.concat(data_jac, axis=0)\n",
    "        heatmap_jac = heatmap_jac.sort_index(axis=0).sort_index(axis=1)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if input_types.raw_value == \"Multiple Query Compounds\" %}\n",
    "if embed_method == 'ReSimNet':\n",
    "    resavg = sns.clustermap(heatmap_avg, annot=True, cmap='OrRd', figsize=(6,6))\n",
    "    resavg.ax_heatmap.set_xlabel('')\n",
    "    resavg.ax_heatmap.set_ylabel('')\n",
    "    resavg.fig.suptitle('Heatmap_ReSimNet avg', y=1, x=0.5);\n",
    "    resavg_filename = 'heatmap_resimnet_avg.png'\n",
    "    plt.savefig(resavg_filename, dpi=300)\n",
    "    plt.close()\n",
    "\n",
    "    resjac = sns.clustermap(heatmap_jac, annot=True, cmap='OrRd', figsize=(6,6))\n",
    "    resjac.ax_heatmap.set_xlabel('')\n",
    "    resjac.ax_heatmap.set_ylabel('')\n",
    "    resjac.fig.suptitle('Heatmap_Jaccard similarity', y=1, x=0.5);\n",
    "    resjac_filename = 'heatmap_jaccard_similarity.png'\n",
    "    plt.savefig(resjac_filename, dpi=300)\n",
    "    plt.close()\n",
    "    \n",
    "    fig, axes = plt.subplots(1, 2, figsize=(12, 6))\n",
    "\n",
    "    axes[0].imshow(plt.imread(resavg_filename))\n",
    "    axes[0].axis('off')\n",
    "\n",
    "    axes[1].imshow(plt.imread(resjac_filename))\n",
    "    axes[1].axis('off')\n",
    "\n",
    "    plt.figtext(0.5, 0.01, f\"figure 2. Heatmap for the Entered Query Compounds\", ha=\"center\", fontsize=10)\n",
    "    plt.show()\n",
    "\n",
    "    display(HTML(f'<a href=\"./heatmap_resimnet_avg.png\" download>Download: Heatmap_ReSimNet avg</a>'))\n",
    "    display(HTML(f'<a href=\"./heatmap_jaccard_similarity.png\" download>Download: Heatmap_Jaccard similarity</a>'))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Drug-Set Enrichment Analysis (DSEA)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drug-Set Enrichment Analysis Results by Drug Enrichr\n",
    "Drug Enrichr is a tool that performs drug set enrichment analysis (DSEA) in terms of targets, mode of actions, side effects, pathways or ontologies. Dr.Emb Appyter performs DSEA by entering the names of the top k compounds into Drug Enrichr. <br> <br>\n",
    "Warning: If the compound names in the library are not common, it may affect the quality of the analysis results. <br> <br>\n",
    "To explore additional drug set results, please click the link provided below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "upset_dict = {}\n",
    "for name in drug_dict.keys():\n",
    "    upset_dict[f\"{name}\"] = list(globals()[f'{name}_result_df']['compound_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for drug, candidates in upset_dict.items():\n",
    "    def drug_enrichr_link(dataset, title = \"Drug Enrichr Overall Results: {}\"):\n",
    "        url = f\"https://maayanlab.cloud/DrugEnrichr/enrich?dataset={dataset}\"\n",
    "        html = \"<a href=\\\"{}\\\" target='_blank'>{}</a>\".format(url, title.format(drug))\n",
    "        return HTML(html)\n",
    "\n",
    "    ENRICHR_URL = 'http://amp.pharm.mssm.edu/DrugEnrichr/addList'\n",
    "    drugs_str = '\\n'.join(candidates)\n",
    "    description = f'{drug} candidates'\n",
    "    payload = {\n",
    "        'list': (None, drugs_str),\n",
    "        'description': (None, description)\n",
    "    }\n",
    "\n",
    "    response = requests.post(ENRICHR_URL, files=payload, verify=False)\n",
    "    if not response.ok:\n",
    "        raise Exception('Error analyzing drug list')\n",
    "\n",
    "    data = json.loads(response.text)\n",
    "    \n",
    "    ENRICHR_URL = 'http://amp.pharm.mssm.edu/DrugEnrichr/view?userListId=%s'\n",
    "    user_list_id = data['userListId']\n",
    "    dataset = data['shortId']\n",
    "    response = requests.get(ENRICHR_URL % user_list_id, verify=False)\n",
    "    if not response.ok:\n",
    "        raise Exception('Error getting drug list')\n",
    "        \n",
    "    data = json.loads(response.text)\n",
    "    \n",
    "    ENRICHR_URL = 'http://amp.pharm.mssm.edu/DrugEnrichr/enrich'\n",
    "    query_string = '?userListId=%s&backgroundType=%s'\n",
    "    drug_set_library = 'Drug_Repurposing_Hub_Mechanism_of_Action'\n",
    "    response = requests.get(\n",
    "        ENRICHR_URL + query_string % (user_list_id, drug_set_library), verify=False\n",
    "    )\n",
    "    if not response.ok:\n",
    "        raise Exception('Error fetching enrichment results')\n",
    "\n",
    "    data = json.loads(response.text)\n",
    "    moa = [item[1] for item in data['Drug_Repurposing_Hub_Mechanism_of_Action']]\n",
    "    p_value = [item[2] for item in data['Drug_Repurposing_Hub_Mechanism_of_Action']]\n",
    "    z_score = [item[3] for item in data['Drug_Repurposing_Hub_Mechanism_of_Action']]\n",
    "    comb_score = [item[4] for item in data['Drug_Repurposing_Hub_Mechanism_of_Action']]\n",
    "\n",
    "    df = pd.DataFrame({'moa': moa, 'p-value': p_value, 'z-score': z_score, 'combined score': comb_score})\n",
    "    styles = [dict(selector=\"caption\", props=[(\"text-align\", \"left\")])] \n",
    "    caption = f\"Table {tb_number}. {drug}'s Mode of Action (MoA) Analysis Results Using Drug Enrichr\"\n",
    "    if df.empty:\n",
    "        caption += \" ※ If you didn't get any results from Drug Enrichr, you need to configure the name of your custom library to a common name.\" \n",
    "    df = df.to_html(classes='styled-table', escape=False)\n",
    "    styled_html = f\"<caption>{caption}</caption>{df}\"\n",
    "    tb_number += 1\n",
    "    display(HTML(styled_html))\n",
    "    display(drug_enrichr_link(dataset))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Acknowledgement\n",
    "The chemical library used in this study was kindly provided by Korea Chemical Bank (http://www.chemicalbank.org/) of Korea Research Institute of Chemical Technology, Sellckchem (https://www.selleckchem.com/) of Selleck Chemicals LLC, and MCE (https://www.medchemexpress.com/) of MedChemExpress LLC.\n",
    "This work was supported by the National Research Foundation of Korea Grant funded by the Korean Government (NRF-2022R1F1A1070111) and the MSIT(Ministry of Science and ICT), Korea, under the ICAN (ICT Challenge and Advanced Network of HRD) program(IITP-2023-RS-2022-00156439) supervised by the IITP (Institute of Information & Communications Technology Planning & Evaluation)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "1. Clarke, D. J. et al. (2021). Appyters: Turning jupyter notebooks into data-driven web apps. Patterns, 2(3).\n",
    "2. Jeon, M. et al. (2019). Resimnet: drug response similarity prediction using siamese neural networks. Bioinformatics, 35(24), 5249–5256.\n",
    "3. Jang, G. et al. (2021). Predicting mechanism of action of novel compounds using compound structure and transcriptomic signature coembedding. Bioinformatics, 37(Supplement_1), i376–i382.\n",
    "4. Rogers, D. and Hahn, M. (2010). Extended-connectivity fingerprints. Journal of chemical information and modeling, 50(5), 742–754.\n",
    "5. Durant, J. L. et al. (2002). Reoptimization of mdl keys for use in drugdiscovery. Journal of chemical information and computer sciences, 42(6), 1273–1280.\n",
    "6. Jaeger, S. et al. (2018). Mol2vec: unsupervised machine learning approach with chemical intuition. Journal of chemical information and modeling, 58(1), 27–35.\n",
    "7. Blay, V. et al. (2022). Macaw: an accessible tool for molecular embedding and inverse molecular design. Journal of chemical information and modeling, 62(15), 3551–3564.\n",
    "8. Johnson, J. et al. (2019). Billion-scale similarity search with GPUs. IEEE Transactions on Big Data, 7(3), 535–547.\n",
    "9. McInnes, L. et al. (2018). UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction. ArXiv e-prints.\n",
    "10. Kuleshov, M. V. et al. (2019). modEnrichr: a suite of gene set enrichment analysis tools for model organisms. Nucleic Acids Research, 47(W1), W183–W190."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dremb",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
